### PR TITLE
MM-15116 Migrates the WebHook.SaveIncoming method to Sync by default

### LINKED ISSUE #10616: [Help Wanted] [MM-15116] Migrate "WebHook.SaveIncomming" to Sync by default
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-15116)

Mattermost is migrating its Store layer to be sync by default, and only use Async when needed and we're looking for contributors to help with that effort. This Help Wanted issue is to migrate the `SaveIncomming` in the `WebHook` store.

The expected way to implement it is, go to the `WebHook` store implementation in the `store/sqlstore/` directory, modify the method `SaveIncomming` to return directly an object from the `model` module, and a `*model.AppError` (removing the `store.Do` wrapper). After that, you must modify the interface defined in `store/store.go` to match with the changes. Then, you should execute `make store-mocks` to rebuild the mocks with the new interface. And finally, modify the rest of the code (tests included) to use the new interface of the function properly.

Example: mattermost/mattermost-server#10613

### LINKED ISSUE #10618: [Help Wanted] [MM-15121] Migrate "WebHook.GetIncomingList" to Sync by default
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-15121)

Mattermost is migrating its Store layer to be sync by default, and only use Async when needed and we're looking for contributors to help with that effort. This Help Wanted issue is to migrate the `GetIncomingList` in the `WebHook` store.

The expected way to implement it is, go to the `WebHook` store implementation in the `store/sqlstore/` directory, modify the method `GetIncomingList` to return directly an object from the `model` module, and a `*model.AppError` (removing the `store.Do` wrapper). After that, you must modify the interface defined in `store/store.go` to match with the changes. Then, you should execute `make store-mocks` to rebuild the mocks with the new interface. And finally, modify the rest of the code (tests included) to use the new interface of the function properly.

Example: mattermost/mattermost-server#10613

### LINKED ISSUE #10655: [Help Wanted] [MM-15198] Migrate "Channel.Get/GetFromMaster" to Sync by default
If you're interested please comment here and come [join our "Contributors" community channel](https://community.mattermost.com/core/channels/tickets) on our daily build server, where you can discuss questions with community members and the Mattermost core team. For technical advice or questions, please  [join our "Developers" community channel](https://community.mattermost.com/core/channels/developers).

New contributors please see our [Developer's Guide](https://developers.mattermost.com/contribute/getting-started/).

----

**Notes**: [Jira ticket](https://mattermost.atlassian.net/browse/MM-15198)

Mattermost is migrating its Store layer to be sync by default, and only use Async when needed and we're looking for contributors to help with that effort. This Help Wanted issue is to migrate the `Get/GetFromMaster` in the `Channel` store.

The expected way to implement it is, go to the `Channel` store implementation in the `store/sqlstore/` directory, modify the method `Get/GetFromMaster` to return directly an object from the `model` module, and a `*model.AppError` (removing the `store.Do` wrapper). After that, you must modify the interface defined in `store/store.go` to match with the changes. Then, you should execute `make store-mocks` to rebuild the mocks with the new interface. And finally, modify the rest of the code (tests included) to use the new interface of the function properly.

Example: mattermost/mattermost-server#10613

### PR DESCRIPTION
**Pull request by andresoro**
#### Summary
Migrates the WebHook.SaveIncoming, WebHook.GetIncomingList, Channel.Get/GetFromMaster methods to sync by default

#### Ticket Link

Fixes #10616
Fixes #10618
Fixes #10655

### CODE REVIEW

File: app/webhook.go [hunk 1]
```diff
@@ -323,13 +323,12 @@ func (a *App) CreateIncomingWebhookForChannel(creatorId string, channel *model.C
 		return nil, model.NewAppError("CreateIncomingWebhookForChannel", "api.incoming_webhook.invalid_username.app_error", nil, "", http.StatusBadRequest)
 	}
 
-	if result := <-a.Srv.Store.Webhook().SaveIncoming(hook); result.Err != nil {
```
Comment (jespino) [inline app/webhook.go]: Prabably here makes sense to return just `return a.Srv.Store.Webhook().SaveIncoming(hook)`

File: app/webhook.go [hunk 2]
```diff
@@ -396,15 +395,17 @@ func (a *App) GetIncomingWebhooksForTeamPage(teamId string, page, perPage int) (
...
+
 	if !*a.Config().ServiceSettings.EnableIncomingWebhooks {
 		return nil, model.NewAppError("GetIncomingWebhooksPage", "api.incoming_webhook.disabled.app_error", nil, "", http.StatusNotImplemented)
 	}
 
-	if result := <-a.Srv.Store.Webhook().GetIncomingList(page*perPage, perPage); result.Err != nil {
-		return nil, result.Err
-	} else {
-		return result.Data.([]*model.IncomingWebhook), nil
+	if webhooks, err = a.Srv.Store.Webhook().GetIncomingList(page*perPage, perPage); err != nil {
```
Comment (jespino) [inline app/webhook.go]: Same here

File: app/webhook.go [hunk 3]
```diff
@@ -396,15 +395,17 @@ func (a *App) GetIncomingWebhooksForTeamPage(teamId string, page, perPage int) (
 }
 
 func (a *App) GetIncomingWebhooksPage(page, perPage int) ([]*model.IncomingWebhook, *model.AppError) {
+	var webhooks []*model.IncomingWebhook
```
Comment (jespino) [inline app/webhook.go]: This lines aren't neede if you use the proposed change.

File: store/storetest/webhook_store.go [hunk 4]
```diff
@@ -121,19 +136,23 @@ func testWebhookStoreGetIncomingList(t *testing.T, ss store.Store) {
...
 }
 
 func testWebhookStoreGetIncomingByTeam(t *testing.T, ss store.Store) {
-	o1 := buildIncomingWebhook()
+	var err *model.AppError
 
-	o1 = (<-ss.Webhook().SaveIncoming(o1)).Data.(*model.IncomingWebhook)
+	o1 := buildIncomingWebhook()
+	o1, err = ss.Webhook().SaveIncoming(o1)
+	if err != nil {
```
Comment (jespino) [inline store/storetest/webhook_store.go]: Not needed, but if you prefer it you can use here `require.Nil(t, err)`.

File: app/webhook.go (hunk 1 above)
Comment (andresoro) [reply app/webhook.go]: Can't believe I didn't notice that. Thanks.

Comment (jespino) [reply app/webhook.go]: No problem :)

**Comment by jespino** [comment]
Please, don't keep adding changes to the same pr, we prefer small prs to review them easily.

**Comment by andresoro** [comment]
My apologies, won’t make the same mistake.

**Comment by jespino** [comment]
Thanks, and thanks for your prs :)

File: app/channel.go [hunk 5]
```diff
@@ -1419,7 +1419,10 @@ func (a *App) postJoinTeamMessage(user *model.User, channel *model.Channel) *mod
 }
 
 func (a *App) LeaveChannel(channelId string, userId string) *model.AppError {
-	sc := a.Srv.Store.Channel().Get(channelId, true)
+	channel, errCh := a.Srv.Store.Channel().Get(channelId, true)
```
Comment (jespino) [inline app/channel.go]: Here we should be using the Async approach, see the example right after your code (`uc` channel creation and the gorouting right after, you have an example in the example PR linked in the issue.

File: app/command.go [hunk 6]
```diff
@@ -218,7 +218,11 @@ func (a *App) tryExecuteCustomCommand(args *model.CommandArgs, trigger string, m
 		return nil, nil, model.NewAppError("ExecuteCommand", "api.command.disabled.app_error", nil, "", http.StatusNotImplemented)
 	}
 
-	chanChan := a.Srv.Store.Channel().Get(args.ChannelId, true)
+	channel, errCh := a.Srv.Store.Channel().Get(args.ChannelId, true)
```
Comment (jespino) [inline app/command.go]: Async approach expected here too

File: app/webhook.go [hunk 7]
```diff
@@ -323,13 +323,9 @@ func (a *App) CreateIncomingWebhookForChannel(creatorId string, channel *model.C
 		return nil, model.NewAppError("CreateIncomingWebhookForChannel", "api.incoming_webhook.invalid_username.app_error", nil, "", http.StatusBadRequest)
 	}
 
-	if result := <-a.Srv.Store.Webhook().SaveIncoming(hook); result.Err != nil {
-		return nil, result.Err
-	} else {
-		return result.Data.(*model.IncomingWebhook), nil
-	}
-}
+	return a.Srv.Store.Webhook().SaveIncoming(hook)
```
Comment (jespino) [inline app/webhook.go]: nit: I would expect this blank line right after the `}` and no blank line between `return` statement and `}`

File: app/webhook.go [hunk 8]
```diff
@@ -396,15 +392,12 @@ func (a *App) GetIncomingWebhooksForTeamPage(teamId string, page, perPage int) (
 }
 
 func (a *App) GetIncomingWebhooksPage(page, perPage int) ([]*model.IncomingWebhook, *model.AppError) {
+
```
Comment (jespino) [inline app/webhook.go]: nit: I would expect no blank line here

**Pull request closed without merging**